File: src/meshplanner/sites/hilltop.py
"""Detect hilltop candidate sites from DEM using morphological dilation.

Uses ``scipy.ndimage.maximum_filter`` to identify local elevation maxima,
computes topographic prominence for each peak via path analysis, and filters
by minimum prominence and inter-peak distance.
"""

import math
from typing import Any

import numpy as np
from scipy.ndimage import maximum_filter, minimum_filter
from scipy.spatial import KDTree

# ── Geo helpers ──────────────────────────────────────────────────────────────


def _haversine_distance(
    lat1: float, lon1: float, lat2: float, lon2: float,
) -> float:
    """Great-circle distance between two points in kilometers."""
    radius = 6371.0
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(dlon / 2) ** 2
    )
    return radius * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def _intermediate_point(
    lat1: float, lon1: float, lat2: float, lon2: float, fraction: float,
) -> tuple[float, float]:
    """Great-circle slerp — returns (lat, lon) at *fraction* along the arc."""
    φ1 = math.radians(lat1)
    λ1 = math.radians(lon1)
    φ2 = math.radians(lat2)
    λ2 = math.radians(lon2)

    # Angular distance
    δ = _haversine_distance(lat1, lon1, lat2, lon2) / 6371.0

    if δ < 1e-12:
        return (lat1, lon1)

    a = math.sin((1 - fraction) * δ) / math.sin(δ)
    b = math.sin(fraction * δ) / math.sin(δ)

    x = a * math.cos(φ1) * math.cos(λ1) + b * math.cos(φ2) * math.cos(λ2)
    y = a * math.cos(φ1) * math.sin(λ1) + b * math.cos(φ2) * math.sin(λ2)
    z = a * math.sin(φ1) + b * math.sin(φ2)

    return (math.degrees(math.atan2(z, math.sqrt(x ** 2 + y ** 2))),
            math.degrees(math.atan2(y, x)))


# ── Bilinear interpolation (nodata-safe) ─────────────────────────────────────


_NODATA_THRESHOLD = -30000


def _bilinear_interpolate(
    dem_array: np.ndarray, col: float, row: float,
) -> float | None:
    """Sample elevation at sub-pixel (col, row).

    Returns ``None`` if outside the DEM or any surrounding pixel is nodata/NaN.
    """
    rows, cols = dem_array.shape
    ci = int(math.floor(col))
    ri = int(math.floor(row))
    fx = col - ci
    fy = row - ri

    if ci < 0 or ri < 0 or ci >= cols or ri >= rows:
        return None

    def _valid(v: float) -> bool:
        return v > _NODATA_THRESHOLD and np.isfinite(v)

    v00 = float(dem_array[ri, ci])
    if not _valid(v00):
        return None

    # Right neighbour
    if fx > 0.0:
        if ci + 1 >= cols:
            return None
        v10 = float(dem_array[ri, ci + 1])
        if not _valid(v10):
            return None
    else:
        v10 = v00

    # Bottom neighbour
    if fy > 0.0:
        if ri + 1 >= rows:
            return None
        v01 = float(dem_array[ri + 1, ci])
        if not _valid(v01):
            return None
    else:
        v01 = v00

    # Diagonal neighbour
    if fx > 0.0 and fy > 0.0:
        if ci + 1 >= cols or ri + 1 >= rows:
            return None
        v11 = float(dem_array[ri + 1, ci + 1])
        if not _valid(v11):
            return None
    elif fx > 0.0:
        v11 = v01
    elif fy > 0.0:
        v11 = v10
    else:
        v11 = v00

    top = v00 + fx * (v10 - v00)
    bottom = v01 + fx * (v11 - v01)
    return top + fy * (bottom - top)


# ── Path-minimum helper (used for prominence saddle search) ──────────────────


def _min_elevation_on_path(
    dem_array: np.ndarray,
    affine: Any,
    lat1: float, lon1: float,
    lat2: float, lon2: float,
    num_samples: int = 50,
) -> float | None:
    """Minimum valid elevation along the great-circle path between two points.

    Returns ``None`` when every sample is invalid.
    """
    min_val = float("inf")
    found = False

    for i in range(num_samples):
        frac = i / (num_samples - 1) if num_samples > 1 else 0.0
        lat, lon = _intermediate_point(lat1, lon1, lat2, lon2, frac)

        col = (lon - affine.c) / affine.a
        row = (lat - affine.f) / affine.e

        elev = _bilinear_interpolate(dem_array, col, row)
        if elev is not None and elev < min_val:
            min_val = elev
            found = True

    return min_val if found else None


# ── Circular footprint ───────────────────────────────────────────────────────


def _make_circular_footprint(radius_px: int) -> np.ndarray:
    """Boolean circular structuring element (diameter = 2*radius_px + 1)."""
    diameter = 2 * radius_px + 1
    fp = np.zeros((diameter, diameter), dtype=bool)
    cx = cy = radius_px
    for i in range(diameter):
        for j in range(diameter):
            if (i - cx) ** 2 + (j - cy) ** 2 <= radius_px ** 2:
                fp[i, j] = True
    return fp


# ── Pixel-size helper ────────────────────────────────────────────────────────


def _compute_pixel_size_km(affine: Any, center_lat: float) -> float:
    """Average pixel size in kilometres, accounting for latitude."""
    lat_rad = math.radians(center_lat)
    km_per_deg_lat = 111.32
    km_per_deg_lon = 111.32 * math.cos(lat_rad)

    px_width_km = abs(affine.a) * km_per_deg_lon
    px_height_km = abs(affine.e) * km_per_deg_lat
    return (px_width_km + px_height_km) / 2.0


# ── Public API ───────────────────────────────────────────────────────────────


def detect_hilltops(
    dem_array: np.ndarray,
    dem_metadata: dict,
    min_prominence_m: float = 50.0,
    min_distance_km: float = 0.5,
) -> list[dict]:
    """Find local elevation maxima in a DEM via morphological dilation.

    The algorithm:

    1.  Mask invalid cells (nodata / NaN).
    2.  Build a circular footprint whose radius (in pixels) is derived from
        *min_distance_km* and the DEM ground-sample distance.
    3.  Apply ``scipy.ndimage.maximum_filter`` — any cell equal to its
        local maximum and strictly greater than the local minimum is a
        candidate peak (the minimum-filter check excludes flat plateaus).
    4.  Convert pixel coordinates to geographic (lat / lon) via the affine
        transform embedded in *dem_metadata*.
    5.  **Prominence**: For each candidate (in descending elevation order),
        find the lowest saddle connecting it to any higher peak by
        sampling elevations along the great-circle arc.  The key saddle
        is the *highest* of those minimum-path values.  Prominence =
        peak elevation − key saddle elevation.
    6.  Filter by *min_prominence_m*.
    7.  Non-maximum suppression: keep only the highest peak within each
        *min_distance_km* exclusion zone.
    8.  Sort by elevation descending.

    Args:
        dem_array:
            2-D float32 elevation array (rows × cols) from
            :func:`~meshplanner.terrain.fetch.fetch_dem_raster`.
        dem_metadata:
            Metadata dict — must contain ``"affine"``
            (a ``rasterio.Affine`` in EPSG:4326).
        min_prominence_m:
            Minimum topographic prominence in metres.  Peaks whose
            prominence is below this threshold are discarded.
            Default ``50.0``.
        min_distance_km:
            Minimum separation between distinct peaks in kilometres.
            Also drives the morphological footprint radius.
            Default ``0.5``.

    Returns:
        List of dicts sorted by descending ``elevation_m``:

        .. code-block:: python

            [
                {
                    "lat": 35.595,
                    "lon": -82.487,
                    "elevation_m": 1542.3,
                    "prominence_m": 312.5,
                },
                ...
            ]

        An empty list is returned when no valid elevation data is present
        or no peaks satisfy the filters.
    """
    dem = np.asarray(dem_array, dtype=np.float32)
    affine = dem_metadata["affine"]

    # ── 1.  Valid-data mask ──────────────────────────────────────────
    valid = np.isfinite(dem) & (dem > _NODATA_THRESHOLD)
    if not valid.any():
        return []

    rows, cols = dem.shape

    # ── 2.  Footprint ────────────────────────────────────────────────
    center_lat = float(affine.f + (rows / 2.0) * affine.e)
    pixel_size_km = _compute_pixel_size_km(affine, center_lat)
    radius_px = max(1, int(round(min_distance_km / pixel_size_km)))
    footprint = _make_circular_footprint(radius_px)

    # ── 3.  Local maxima via morphological dilation ──────────────────
    # Manually pad the array to avoid scipy boundary bugs that occur
    # when the footprint is much larger than the input array.  We use
    # -inf / +inf as pad values so that out-of-bounds cells never
    # become candidates for the max / min filters.
    pad = radius_px  # footprint radius in pixels
    dem_filled = np.where(valid, dem, -np.inf).astype(np.float32)
    dem_padded = np.pad(dem_filled, pad, mode="constant", constant_values=-np.inf)

    dem_max_padded = maximum_filter(dem_padded, footprint=footprint)
    # Crop the padded result back to the original DEM shape.
    dem_max = dem_max_padded[pad:-pad, pad:-pad]

    # For the min-filter we use +inf as the sentinel so out-of-bounds
    # cells never become the neighbourhood minimum.
    dem_min_input = np.where(valid, dem, np.inf).astype(np.float32)
    dem_min_padded = np.pad(dem_min_input, pad, mode="constant", constant_values=np.inf)
    dem_min_padded_ft = minimum_filter(dem_min_padded, footprint=footprint)
    dem_min = dem_min_padded_ft[pad:-pad, pad:-pad]

    # A true local maximum equals the max-filter AND is strictly larger
    # than the min-filter (excludes flat plateaus).
    is_peak = (dem_filled == dem_max) & (dem_filled > dem_min) & valid

    # Exclude cells too close to the DEM edge — the padded boundary
    # can create false-positive peaks where the neighbourhood is only
    # partially valid.
    if pad < min(rows, cols) // 2:
        is_peak[:pad, :] = False
        is_peak[-pad:, :] = False
        is_peak[:, :pad] = False
        is_peak[:, -pad:] = False

    peak_rows, peak_cols = np.where(is_peak)
    if len(peak_rows) == 0:
        return []

    # ── 4.  Pixel → geographic ──────────────────────────────────────
    peak_lons = affine.c + peak_cols * affine.a
    peak_lats = affine.f + peak_rows * affine.e
    peak_elvs = dem[peak_rows, peak_cols]

    peaks: list[dict[str, Any]] = [
        {
            "row": int(r), "col": int(c),
            "lat": float(peak_lats[i]), "lon": float(peak_lons[i]),
            "elevation": float(peak_elvs[i]),
        }
        for i, (r, c) in enumerate(zip(peak_rows, peak_cols))
    ]

    # ── 5.  Sort by elevation descending ─────────────────────────────
    peaks.sort(key=lambda p: p["elevation"], reverse=True)

    # ── 6.  Prominence computation ───────────────────────────────────
    # The key saddle for peak P is the *highest* of the lowest points
    # on all paths from P to a *higher* peak.  For the overall highest
    # peak we approximate the key saddle by checking paths to the DEM
    # boundary (its "island parent").
    #
    # Performance: checking *all* higher peaks is O(n²) and becomes
    # impractical for hundreds of peaks.  We use a KD-tree to find
    # nearby higher peaks and limit the search to the nearest 30.
    # For most topographic settings the key saddle lies close to the
    # peak, so this approximation is accurate.
    #
    # Boundary points used for the highest-peak case.
    boundary_pts = [
        (0, 0),
        (0, cols - 1),
        (rows - 1, 0),
        (rows - 1, cols - 1),
        (rows // 2, 0),
        (rows // 2, cols - 1),
        (0, cols // 2),
        (rows - 1, cols // 2),
    ]

    # Build KD-tree over (lon, lat) for fast spatial queries.
    peak_lon_lat = np.array([[p["lon"], p["lat"]] for p in peaks], dtype=np.float64)
    tree = KDTree(peak_lon_lat)

    for i, peak in enumerate(peaks):
        if i == 0:
            # ── Highest peak — check paths to DEM boundary ──────
            saddle_vals: list[float] = []
            for br, bc in boundary_pts:
                flat2 = float(affine.f + br * affine.e)
                lon2 = float(affine.c + bc * affine.a)
                min_e = _min_elevation_on_path(
                    dem, affine,
                    peak["lat"], peak["lon"], flat2, lon2,
                    num_samples=100,
                )
                if min_e is not None:
                    saddle_vals.append(min_e)

            if saddle_vals and peak["elevation"] > max(saddle_vals):
                prominence = peak["elevation"] - max(saddle_vals)
            else:
                prominence = peak["elevation"]
        else:
            # ── Lower peaks — check nearby higher peaks via KD-tree ──
            # Search for the 30 nearest neighbours (which includes the
            # peak itself at distance 0).  We increase `k` iteratively
            # until we find at least one higher peak.
            k = min(31, len(peaks))
            distances, indices = tree.query(
                peak_lon_lat[i], k=k, p=2,
            )
            # Indices[0] is the peak itself; skip it.
            higher_indices = [idx for idx in indices if idx < i]

            # If none of the nearest 30 are higher, fall back to
            # checking ALL higher peaks (this is rare for large N
            # but can happen for the second-highest peak when peaks
            # are widely spaced).
            if not higher_indices:
                higher_indices = list(range(i))

            saddle_candidates: list[float] = []
            for j in higher_indices:
                higher = peaks[j]
                min_on_path = _min_elevation_on_path(
                    dem, affine,
                    peak["lat"], peak["lon"],
                    higher["lat"], higher["lon"],
                    num_samples=50,
                )
                if min_on_path is not None:
                    saddle_candidates.append(min_on_path)

            if saddle_candidates:
                max_saddle = max(saddle_candidates)
                prominence = peak["elevation"] - max_saddle
            else:
                prominence = peak["elevation"]

        peak["prominence_m"] = max(0.0, prominence)

    # ── 7.  Filter by prominence ─────────────────────────────────────
    peaks = [p for p in peaks if p["prominence_m"] >= min_prominence_m]

    # ── 8.  Non-maximum suppression by min_distance_km ───────────────
    kept: list[dict[str, Any]] = []
    for peak in peaks:
        too_close = any(
            _haversine_distance(peak["lat"], peak["lon"],
                                k["lat"], k["lon"])
            < min_distance_km
            for k in kept
        )
        if not too_close:
            kept.append(peak)

    # ── 9.  Return clean result dicts ────────────────────────────────
    return [
        {
            "lat": p["lat"],
            "lon": p["lon"],
            "elevation_m": p["elevation"],
            "prominence_m": p["prominence_m"],
        }
        for p in kept
    ]

File: meshplanner-app/src/lib/sites/hilltop.ts
/**
 * Detect hilltop candidate sites from DEM data.
 *
 * Port of Python's `meshplanner.sites.hilltop` — no scipy/numpy dependency.
 * Uses manual sliding-window morphological filters instead of
 * ``scipy.ndimage.maximum_filter``, making it suitable for browser
 * environments.
 *
 * Algorithm:
 * 1. Mask invalid cells (nodata / NaN).
 * 2. Build a circular footprint from *minDistanceKm* and the DEM resolution.
 * 3. Apply maximum / minimum morphological filters to identify local maxima
 *    that are strictly higher than the neighbourhood minimum (excludes flat
 *    plateaus).
 * 4. Convert pixel coordinates to (lat, lon) via the affine transform.
 * 5. **Prominence**: For each candidate (descending elevation), find the
 *    lowest saddle connecting it to any higher peak by sampling elevations
 *    along the great-circle arc. The key saddle is the *highest* of those
 *    minimum-path values. Prominence = peak elevation − key saddle
 *    elevation.
 * 6. Filter by *minProminenceM*.
 * 7. Non-maximum suppression: keep only the highest peak within each
 *    *minDistanceKm* exclusion zone.
 * 8. Sort by elevation descending.
 */

import { Affine } from "@/lib/math/affine"

// ── Constants ─────────────────────────────────────────────────────────────────

/** Elevations below this threshold are treated as invalid (nodata). */
const NODATA_THRESHOLD = -30000

/** Earth mean radius in kilometres (WGS‑84). */
const EARTH_RADIUS_KM = 6371.0

// ── Types ─────────────────────────────────────────────────────────────────────

export interface HilltopResult {
  /** Geographic latitude (EPSG:4326). */
  lat: number
  /** Geographic longitude (EPSG:4326). */
  lon: number
  /** Elevation above sea level in metres. */
  elevation_m: number
  /** Topographic prominence in metres. */
  prominence_m: number
}

// ── Internal peak representation ──────────────────────────────────────────────

interface Peak {
  row: number
  col: number
  lat: number
  lon: number
  elevation: number
  /** Computed topographic prominence (set during step 5). */
  prominence_m: number
}

// ── Geo helpers ───────────────────────────────────────────────────────────────

/**
 * Great-circle distance between two points in kilometres (Haversine formula).
 */
function haversineDistance(
  lat1: number, lon1: number,
  lat2: number, lon2: number,
): number {
  const dlat = (lat2 - lat1) * (Math.PI / 180)
  const dlon = (lon2 - lon1) * (Math.PI / 180)
  const a =
    Math.sin(dlat / 2) ** 2 +
    Math.cos(lat1 * (Math.PI / 180)) *
      Math.cos(lat2 * (Math.PI / 180)) *
      Math.sin(dlon / 2) ** 2
  return EARTH_RADIUS_KM * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
}

/**
 * Great-circle spherical linear interpolation (slerp).
 * Returns (lat, lon) at *fraction* (0‑1) along the arc from point 1 to 2.
 */
function intermediatePoint(
  lat1: number, lon1: number,
  lat2: number, lon2: number,
  fraction: number,
): [number, number] {
  const φ1 = lat1 * (Math.PI / 180)
  const λ1 = lon1 * (Math.PI / 180)
  const φ2 = lat2 * (Math.PI / 180)
  const λ2 = lon2 * (Math.PI / 180)

  // Angular distance in radians
  const δ = haversineDistance(lat1, lon1, lat2, lon2) / EARTH_RADIUS_KM

  if (δ < 1e-12) return [lat1, lon1]

  const a = Math.sin((1 - fraction) * δ) / Math.sin(δ)
  const b = Math.sin(fraction * δ) / Math.sin(δ)

  const x = a * Math.cos(φ1) * Math.cos(λ1) + b * Math.cos(φ2) * Math.cos(λ2)
  const y = a * Math.cos(φ1) * Math.sin(λ1) + b * Math.cos(φ2) * Math.sin(λ2)
  const z = a * Math.sin(φ1) + b * Math.sin(φ2)

  return [
    Math.atan2(z, Math.sqrt(x * x + y * y)) * (180 / Math.PI),
    Math.atan2(y, x) * (180 / Math.PI),
  ]
}

// ── Bilinear interpolation (nodata‑safe) ──────────────────────────────────────

/**
 * Sample elevation at sub‑pixel (col, row) using bilinear interpolation.
 *
 * Returns ``null`` if outside the DEM or any surrounding pixel is
 * nodata/NaN.
 */
function bilinearInterpolate(
  dem: Float32Array,
  width: number,
  height: number,
  col: number,
  row: number,
): number | null {
  const ci = Math.floor(col)
  const ri = Math.floor(row)
  const fx = col - ci
  const fy = row - ri

  if (ci < 0 || ri < 0 || ci >= width || ri >= height) return null

  const isValid = (v: number) => v > NODATA_THRESHOLD && Number.isFinite(v)

  const v00 = dem[ri * width + ci]!
  if (!isValid(v00)) return null

  // Right neighbour
  let v10: number
  if (fx > 0) {
    if (ci + 1 >= width) return null
    v10 = dem[ri * width + ci + 1]!
    if (!isValid(v10)) return null
  } else {
    v10 = v00
  }

  // Bottom neighbour
  let v01: number
  if (fy > 0) {
    if (ri + 1 >= height) return null
    v01 = dem[(ri + 1) * width + ci]!
    if (!isValid(v01)) return null
  } else {
    v01 = v00
  }

  // Diagonal neighbour
  let v11: number
  if (fx > 0 && fy > 0) {
    if (ci + 1 >= width || ri + 1 >= height) return null
    v11 = dem[(ri + 1) * width + ci + 1]!
    if (!isValid(v11)) return null
  } else if (fx > 0) {
    v11 = v01
  } else if (fy > 0) {
    v11 = v10
  } else {
    v11 = v00
  }

  const top = v00 + fx * (v10 - v00)
  const bottom = v01 + fx * (v11 - v01)
  return top + fy * (bottom - top)
}

// ── Path‑minimum helper (used for prominence saddle search) ──────────────────

/**
 * Minimum valid elevation along the great‑circle path between two points.
 *
 * Returns ``null`` when every sample is invalid.
 */
function minElevationOnPath(
  dem: Float32Array,
  width: number,
  height: number,
  affine: Affine,
  lat1: number, lon1: number,
  lat2: number, lon2: number,
  numSamples = 50,
): number | null {
  let minVal = Infinity
  let found = false

  for (let i = 0; i < numSamples; i++) {
    const frac = numSamples > 1 ? i / (numSamples - 1) : 0
    const [lat, lon] = intermediatePoint(lat1, lon1, lat2, lon2, frac)

    // Convert geographic → pixel using the affine inverse
    const col = (lon - affine.c) / affine.a
    const row = (lat - affine.f) / affine.e

    const elev = bilinearInterpolate(dem, width, height, col, row)
    if (elev !== null && elev < minVal) {
      minVal = elev
      found = true
    }
  }

  return found ? minVal : null
}

// ── Circular footprint ────────────────────────────────────────────────────────

/**
 * Pre‑compute the (dr, dc) offsets that fall within a circle of *radiusPx*.
 * This avoids repeatedly iterating over empty cells during the sliding window.
 */
function makeCircularFootprintOffsets(radiusPx: number): [number, number][] {
  const radiusSq = radiusPx * radiusPx
  const offsets: [number, number][] = []
  for (let dr = -radiusPx; dr <= radiusPx; dr++) {
    for (let dc = -radiusPx; dc <= radiusPx; dc++) {
      if (dr * dr + dc * dc <= radiusSq) {
        offsets.push([dr, dc])
      }
    }
  }
  return offsets
}

/**
 * Compute the average pixel size in kilometres at a given latitude.
 * Accounts for the varying length of a degree of longitude.
 */
function computePixelSizeKm(affine: Affine, centerLat: number): number {
  const latRad = centerLat * (Math.PI / 180)
  const kmPerDegLat = 111.32
  const kmPerDegLon = 111.32 * Math.cos(latRad)

  const pxWidthKm = Math.abs(affine.a) * kmPerDegLon
  const pxHeightKm = Math.abs(affine.e) * kmPerDegLat
  return (pxWidthKm + pxHeightKm) / 2
}

// ── Morphological filters (manual sliding window — no scipy) ──────────────────

/**
 * Manual maximum filter over a circular neighbourhood.
 *
 * For each cell, finds the maximum value among neighbours within the
 * pre‑computed *offsets* (from ``makeCircularFootprintOffsets``).
 */
function maximumFilter(
  data: Float32Array,
  width: number,
  height: number,
  offsets: [number, number][],
): Float32Array {
  const result = new Float32Array(width * height)
  for (let r = 0; r < height; r++) {
    for (let c = 0; c < width; c++) {
      let maxVal = -Infinity
      for (const [dr, dc] of offsets) {
        const nr = r + dr
        const nc = c + dc
        if (nr >= 0 && nr < height && nc >= 0 && nc < width) {
          const v = data[nr * width + nc]!
          if (v > maxVal) maxVal = v
        }
      }
      result[r * width + c] = maxVal
    }
  }
  return result
}

/**
 * Manual minimum filter over a circular neighbourhood.
 */
function minimumFilter(
  data: Float32Array,
  width: number,
  height: number,
  offsets: [number, number][],
): Float32Array {
  const result = new Float32Array(width * height)
  for (let r = 0; r < height; r++) {
    for (let c = 0; c < width; c++) {
      let minVal = Infinity
      for (const [dr, dc] of offsets) {
        const nr = r + dr
        const nc = c + dc
        if (nr >= 0 && nr < height && nc >= 0 && nc < width) {
          const v = data[nr * width + nc]!
          if (v < minVal) minVal = v
        }
      }
      result[r * width + c] = minVal
    }
  }
  return result
}

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Find local elevation maxima in a DEM.
 *
 * @param demData   1‑D elevation array in row‑major order (Float32Array).
 * @param demWidth  Number of columns (pixels per row).
 * @param demHeight Number of rows.
 * @param affine    Affine transform (EPSG:4326) mapping pixel → geographic.
 * @param minProminenceM  Minimum topographic prominence in metres (default 50).
 * @param minDistanceKm   Minimum separation between distinct peaks in km
 *                        (default 0.5).  Also drives the footprint radius.
 *
 * @returns Array of {@link HilltopResult} sorted by descending elevation.
 *          Empty when there are no valid data or no peaks satisfy the filters.
 */
export function detectHilltops(
  demData: Float32Array,
  demWidth: number,
  demHeight: number,
  affine: Affine,
  minProminenceM = 50.0,
  minDistanceKm = 0.5,
): HilltopResult[] {
  const rows = demHeight
  const cols = demWidth

  // ── 1.  Valid‑data mask ──────────────────────────────────────────
  const valid = new Uint8Array(demData.length)
  let anyValid = false
  for (let i = 0; i < demData.length; i++) {
    const v = demData[i]!
    if (Number.isFinite(v) && v > NODATA_THRESHOLD) {
      valid[i] = 1
      anyValid = true
    }
  }
  if (!anyValid) return []

  // ── 2.  Footprint ────────────────────────────────────────────────
  const centerLat = affine.f + (rows / 2) * affine.e
  const pixelSizeKm = computePixelSizeKm(affine, centerLat)
  const radiusPx = Math.max(1, Math.round(minDistanceKm / pixelSizeKm))
  const offsets = makeCircularFootprintOffsets(radiusPx)
  const pad = radiusPx

  // ── 3.  Fill invalid cells with sentinel values for filtering ────
  // Invalid cells → −Inf for max filter (never picked as maximum),
  //                +Inf for min filter (never picked as minimum).
  const demFilled = new Float32Array(demData)
  const demMinInput = new Float32Array(demData)
  for (let i = 0; i < demData.length; i++) {
    if (!valid[i]) {
      demFilled[i] = -Infinity
      demMinInput[i] = Infinity
    }
  }

  // ── 4.  Maximum / minimum morphological filters ──────────────────
  const demMax = maximumFilter(demFilled, cols, rows, offsets)
  const demMin = minimumFilter(demMinInput, cols, rows, offsets)

  // ── 5.  Detect local maxima ──────────────────────────────────────
  // A true peak:
  //   · equals its neighbourhood maximum,
  //   · is strictly greater than the neighbourhood minimum (→ excludes flat
  //     plateaus), and
  //   · is a valid cell.
  // Cells within `pad` pixels of the DEM edge are excluded because the
  // truncated neighbourhood creates false positives.  Skip the exclusion
  // when the footprint is larger than half the grid (nothing would pass).
  const edgePad = pad < Math.min(rows, cols) / 2 ? pad : 0
  const isPeak = new Uint8Array(demData.length)
  for (let r = edgePad; r < rows - edgePad; r++) {
    for (let c = edgePad; c < cols - edgePad; c++) {
      const idx = r * cols + c
      if (
        valid[idx] &&
        demFilled[idx]! === demMax[idx]! &&
        demFilled[idx]! > demMin[idx]!
      ) {
        isPeak[idx] = 1
      }
    }
  }

  // Collect peak cells
  const peaks: Peak[] = []
  for (let r = edgePad; r < rows - edgePad; r++) {
    for (let c = edgePad; c < cols - edgePad; c++) {
      const idx = r * cols + c
      if (isPeak[idx]) {
        const [lon, lat] = affine.pixelToGeo(c, r)
        peaks.push({
          row: r,
          col: c,
          lat,
          lon,
          elevation: demData[idx]!,
          prominence_m: 0, // placeholder, computed below
        })
      }
    }
  }

  if (peaks.length === 0) return []

  // ── 6.  Sort by elevation descending ─────────────────────────────
  peaks.sort((a, b) => b.elevation - a.elevation)

  // ── 7.  Prominence computation ───────────────────────────────────
  // The key saddle for peak P is the highest of the lowest points on
  // all paths from P to a *higher* peak.  For the overall highest peak
  // we approximate the key saddle by checking paths to the DEM boundary
  // (its "island parent").
  //
  // Performance: checking all higher peaks is O(n²) and becomes
  // impractical.  We check the 30 nearest higher peaks by spatial
  // proximity.  For most topographic settings the key saddle lies close
  // to the peak, so this approximation is accurate.

  // Boundary points used for the highest‑peak case.
  const boundaryPts: [number, number][] = [
    [0, 0],
    [0, cols - 1],
    [rows - 1, 0],
    [rows - 1, cols - 1],
    [rows >> 1, 0],
    [rows >> 1, cols - 1],
    [0, cols >> 1],
    [rows - 1, cols >> 1],
  ]

  for (let i = 0; i < peaks.length; i++) {
    const peak = peaks[i]!

    if (i === 0) {
      // ── Highest peak — check paths to DEM boundary ──────
      const saddleVals: number[] = []
      for (const [br, bc] of boundaryPts) {
        const lat2 = affine.f + br * affine.e
        const lon2 = affine.c + bc * affine.a
        const minE = minElevationOnPath(
          demData, cols, rows, affine,
          peak.lat, peak.lon, lat2, lon2,
          100,
        )
        if (minE !== null) saddleVals.push(minE)
      }

      if (saddleVals.length > 0 && peak.elevation > Math.max(...saddleVals)) {
        peak.prominence_m = peak.elevation - Math.max(...saddleVals)
      } else {
        peak.prominence_m = peak.elevation
      }
    } else {
      // ── Lower peaks — check nearby higher peaks ─────────
      // Compute squared‑distance proximity to find the 30 nearest
      // neighbours (substitutes for scipy.spatial.KDTree which is
      // not available in the browser).
      const MAX_K = 31
      const distances: { idx: number; distSq: number }[] = []
      for (let j = 0; j < peaks.length; j++) {
        const p = peaks[j]!
        distances.push({
          idx: j,
          distSq: (p.lat - peak.lat) ** 2 + (p.lon - peak.lon) ** 2,
        })
      }
      // Sort by proximity (stable for equal distances — not critical).
      distances.sort((a, b) => a.distSq - b.distSq)

      const higherIndices: number[] = []
      const limit = Math.min(MAX_K, distances.length)
      for (let k = 0; k < limit; k++) {
        if (distances[k]!.idx < i) {
          higherIndices.push(distances[k]!.idx)
        }
      }

      // If none of the nearest 30 are higher, fall back to
      // checking ALL higher peaks (rare for large N, but can
      // happen for the second‑highest peak when widely spaced).
      if (higherIndices.length === 0) {
        for (let j = 0; j < i; j++) higherIndices.push(j)
      }

      const saddleCandidates: number[] = []
      for (const j of higherIndices) {
        const higher = peaks[j]!
        const minOnPath = minElevationOnPath(
          demData, cols, rows, affine,
          peak.lat, peak.lon,
          higher.lat, higher.lon,
          50,
        )
        if (minOnPath !== null) saddleCandidates.push(minOnPath)
      }

      if (saddleCandidates.length > 0) {
        peak.prominence_m = peak.elevation - Math.max(...saddleCandidates)
      } else {
        peak.prominence_m = peak.elevation
      }
    }

    peak.prominence_m = Math.max(0, peak.prominence_m)
  }

  // ── 8.  Filter by minimum prominence ────────────────────────────
  const prominentPeaks = peaks.filter(p => p.prominence_m >= minProminenceM)

  // ── 9.  Non‑maximum suppression by minimum distance ──────────────
  const kept: Peak[] = []
  for (const p of prominentPeaks) {
    const tooClose = kept.some(
      k => haversineDistance(p.lat, p.lon, k.lat, k.lon) < minDistanceKm,
    )
    if (!tooClose) kept.push(p)
  }

  // ── 10.  Return clean result dicts ───────────────────────────────
  return kept.map(p => ({
    lat: p.lat,
    lon: p.lon,
    elevation_m: p.elevation,
    prominence_m: p.prominence_m,
  }))
}
